src/libsystemd/sd-event/sd-event.c

```diff
@@ -2242,188 +2242,197 @@ _public_ int sd_event_source_get_enabled(sd_event_source *s, int *m) {
+                source_time_prioq_reshuffle(s, s->type);
+                break;
 
-                        if (EVENT_SOURCE_WATCH_PIDFD(s))
-                                source_child_pidfd_unregister(s);
-                        else
-                                event_gc_signal_data(s->event, &s->priority, SIGCHLD);
+        case SOURCE_SIGNAL:
+                event_gc_signal_data(s->event, &s->priority, s->signal.sig);
+                break;
 
-                        break;
+        case SOURCE_CHILD:
+                assert(s->event->n_enabled_child_sources > 0);
+                s->event->n_enabled_child_sources--;
 
-                case SOURCE_EXIT:
-                        s->enabled = m;
-                        prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
-                        break;
+                if (EVENT_SOURCE_WATCH_PIDFD(s))
+                        source_child_pidfd_unregister(s);
+                else
+                        event_gc_signal_data(s->event, &s->priority, SIGCHLD);
 
-                case SOURCE_DEFER:
-                case SOURCE_POST:
-                case SOURCE_INOTIFY:
-                        s->enabled = m;
-                        break;
+                break;
 
-                default:
-                        assert_not_reached("Wut? I shouldn't exist.");
-                }
+        case SOURCE_EXIT:
+                prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
+                break;
 
-        } else {
+        case SOURCE_DEFER:
+        case SOURCE_POST:
+        case SOURCE_INOTIFY:
+                break;
 
-                /* Unset the pending flag when this event source is enabled */
-                if (s->enabled == SD_EVENT_OFF && !IN_SET(s->type, SOURCE_DEFER, SOURCE_EXIT)) {
-                        r = source_set_pending(s, false);
-                        if (r < 0)
-                                return r;
-                }
+        default:
+                assert_not_reached("Wut? I shouldn't exist.");
+        }
 
-                switch (s->type) {
+        source_prioq_reshuffle(s);
 
-                case SOURCE_IO:
-                        r = source_io_register(s, m, s->io.events);
-                        if (r < 0)
-                                return r;
+        return 0;
+}
 
-                        s->enabled = m;
-                        break;
+static int source_enable(sd_event_source *s, int m) {
+        int r;
 
-                case SOURCE_TIME_REALTIME:
-                case SOURCE_TIME_BOOTTIME:
-                case SOURCE_TIME_MONOTONIC:
-                case SOURCE_TIME_REALTIME_ALARM:
-                case SOURCE_TIME_BOOTTIME_ALARM: {
-                        struct clock_data *d;
+        assert(s);
+        assert(m != SD_EVENT_OFF);
+
+        /* Unset the pending flag when this event source is enabled */
+        if (s->enabled == SD_EVENT_OFF && !IN_SET(s->type, SOURCE_DEFER, SOURCE_EXIT)) {
```

Comment:
hmm, you updated s->enabled already in the caller before calling source_enable(), which means `s->enabled` is already identical to `m` here, and hence not `SD_EVENT_OFF`.

Reply:
I've changed the update order in the caller. test-event is still passing so hopefully that change doesn't have any negative side effects.